activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -321,7 +320,7 @@ def shard
       end
 
       def schema_cache
-        @schema_cache ||= BoundSchemaReflection.new(@pool.schema_reflection, self)
+        @pool.schema_cache || (@schema_cache ||= BoundSchemaReflection.for_lone_connection(@pool.schema_reflection, self))
```

Comment:
I think it's fine for this to just return the pool's bound schema cache. The significance of the binding was to maintain a caller's ability to grab a `schema_cache` and then ask it questions without having to worry about passing in the connection each time, rather than to strictly enforce the use of the individual connection.

The only situation that comes to mind would be if a caller is inside a transaction, has just performed DDL, and then makes [cache-missing] inquiries: if this connection was manually checked out, and so is not their pinned connection [pinned by the open transaction, if nothing else], then using the "wrong" connection would give a different result. But then, deliberately populating the schema cache with state that's only visible to a single transaction seems like a questionable idea anyway.

(I don't _object_ to `for_lone_connection` -- just wondered about whether it would be a short term deprecation-guiderail or a long term API, and then ended up at this feeling that it's probably an ignorable behavioural distinction.)

Reply:
Oh, I suppose the risk would be that the current thread doesn't have a pinned connection for this pool at all, and so we'd be requesting something new -- at least inflating the total number of concurrent connections required, and possibly running afoul of either local or server-side connection limits. 

So yeah we should probably do this despite the minor awkwardness. 👍🏻

Reply:
So that `for_lone_connection` is for Connections that don't belong to a pool, so `@pool` is `NullPool`. 

This is a case that is tested but I'm not sure how much it really happens in practice.